department-of-veterans-affairs/vets-api
# Api 41367 replace bgs_ext special issue updater

## Summary
* Adds feature flag to redirect `find_contentions_by_ptcpnt_id` & `manage_contentions` to use local_bgs instead of bgs_ext
*  Adds definition for `contention_service` and contention_service file
    * Adds test for `contention_service`
* Updates `special_issues_updater` to use local_bgs if Flipper enabled
* Updates `special_issue_updater_spec.rb` tests for flipper enabled/disabled scenarios
* Adds VCR cassette for contention `find_contentions_by_ptcpnt_id.yml`

## Related issue(s)
[API-41367](https://jira.devops.va.gov/browse/API-41367)

## Testing done

#### Testing Notes
* Enable Flipper: `Flipper.enable(:claims_api_special_issues_updater_uses_local_bgs)`
* Submit a claim in v1 that would get flagged for special issues
    *  One thing to do is add confinements
    ```
            "confinements":[
               {
                  "confinementBeginDate":"1989-01-01",
                  "confinementEndDate":"1989-12-01"
               }
            ],
   ```
    * Then add `"specialIssues": ["POW"]` to a disability 
*  think the easiest way to test the results it to debug out in `local_bgs` after line 307.  Just check the response value there and make sure a 200 cam eback.  You can also check the body getting sent which for manage contentions can be useful to see (I would suggest `pp body.to_s` to view that).  Check the `action` first before printing out just to make sure you know what call actually is running, `find_contentions_by_ptcpnt_id` will likely be massive.
* To just verify it is being used, with the feature flag enabled, run `rails c`
 ```
@ssl_verify_mode =
if Settings.bgs.ssl_verify_mode == 'none'
OpenSSL::SSL::VERIFY_NONE
else
OpenSSL::SSL::VERIFY_PEER
end

ClaimsApi::LocalBGS.new(external_uid:'a', external_key:'b').fetch_namespace(Faraday::Connection.new(ssl: { verify_mode: @ssl_verify_mode }), "ContentionService/ContentionService")
```
* Should get `=> "http://services.mapd.benefits.vba.va.gov/"`

## Screenshots

## What areas of the site does it impact?
    modified: config/features.yml
    modified: modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb
    modified: modules/claims_api/app/sidekiq/claims_api/special_issue_updater.rb
    new: modules/claims_api/lib/bgs_service/contention_service.rb
    new: modules/claims_api/spec/lib/claims_api/bgs/contention_service_spec.rb
    modified: modules/claims_api/spec/sidekiq/special_issue_updater_spec.rb
    new: spec/support/vcr_cassettes/claims_api/bgs/contention/find_contentions_by_ptcpnt_id.yml

## Acceptance criteria

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_
